File: app/models/ebean/Invite.java
package models.ebean;

import com.avaje.ebean.Model;
import org.jetbrains.annotations.NotNull;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Entity
public class Invite extends Model{

    @Id
    private UUID id;
    @NotNull
    @ManyToOne
    private Player sender;
    @NotNull
    @ManyToOne
    private Player receiver;
    @NotNull
    @ManyToOne
    private Team team;

    private boolean answered;
    private String requestType;

    private static Model.Finder<UUID, Invite> finder = new Model.Finder<>(Invite.class);

    public Invite(UUID id, @NotNull Player sender, @NotNull Player receiver, @NotNull Team team, boolean answered, String requestType) {
        this.id = id;
        this.sender = sender;
        this.receiver = receiver;
        this.team = team;
        this.answered = answered;
        this.requestType = requestType;
    }

    public static Optional<Invite> getById(UUID id) {
        Invite team = finder.where().eq("id", id).findUnique();
        if(team != null) {
            return  Optional.of(team);
        } else {
            return Optional.empty();
        }
    }
    
    public static List<Invite> getReceivedInvites(UUID receiverId) {
        return finder.where().eq("receiver_id", receiverId).findList();
    }

    public static List<Invite> checkJoinRequests(UUID senderId, UUID teamId) {
        return finder.where().eq("sender_id", senderId).eq("team_id", teamId)
                .eq("request_type", "Join").findList();
    }

    public static List<Invite> getPlayerNotifications(UUID receiverId) {
        return finder.where().eq("receiver_id", receiverId).eq("answered", false).findList();
    }


    public UUID getId() {
        return id;
    }

    @NotNull
    public Player getSender() {
        return sender;
    }

    @NotNull
    public Player getReceiver() {
        return receiver;
    }

    public boolean isAnswered() {
        return answered;
    }

    public Team getTeam() {
        return team;
    }

    public String getType() {
        return requestType;
    }
}

File: app/models/domain/invite/Invite.scala
package models.domain.invite

import java.util.UUID

import models.dao.InviteDAO
import models.domain.player.Player
import models.domain.team.Team
import play.api.libs.json.{Json, OFormat}
import models.ebean.{Invite => EInvite}

case class Invite(id: UUID, sender: Player, receiver: Player, team: Team, answered: Boolean, requestType: String)

object Invite {
  implicit val format: OFormat[Invite] = Json.format
  
  def apply(eInvite: EInvite): Invite = {
    Invite(eInvite.getId, Player(eInvite.getSender), Player(eInvite.getReceiver), Team(eInvite.getTeam), eInvite.isAnswered, eInvite.getType)
  }
  
  def save(sender: Player, receiver: Player, team: Team, requestType: String): Invite = {
    InviteDAO.save(Invite(UUID.randomUUID(), sender, receiver, team, false, requestType))
  }

  def checkJoinRequests(senderId: UUID, teamId: UUID): List[Invite] = {
    InviteDAO.checkJoinRequests(senderId, teamId)
  }

  def getPlayerNotifications(receiver: UUID): List[Invite] = {
    InviteDAO.getPlayerNotifications(receiver)
  }

  def getById(id: UUID): Option[Invite] = {
    InviteDAO.getById(id)
  }

  def accept(invite: Invite) = {
    InviteDAO.accept(invite)
  }

  def acceptTeamInvite(invite: Invite) = {
    InviteDAO.acceptTeamInvite(invite)
  }


  def reject(invite: Invite) = {
    InviteDAO.reject(invite)
  }
}
